acl2_kernel/kernel.py

```diff
@@ -51,8 +52,12 @@ def do_execute(self, code, silent, store_history=True, user_expressions=None, al
             }
         interrupted = False
         try:
-            cmd = re.sub(r'[\r\n]|;[^\r\n]*[\r\n]+', ' ', code.strip())
+            cmd = re.sub(r';.*$', '', code, flags=re.MULTILINE)
+            num_cmds = len(regex.findall(r'^[ \t]*:.*$|\((?>[^()]|(?R))*\)', cmd, regex.MULTILINE))
+            cmd = re.sub(r'[\r\n]', ' ', cmd.strip())
```

Comment:
Could you add the example of text matched by these regular expressions as comments because it is hard to know what you express for each lines

Reply:
Those regex lines were tested and work on these 3 code blocks:

1.
```lisp
:pe subsetp
;:pl (+ 1 2)
```

2.
```lisp
;:pe (+ 1 2)
:pl subsetp
```

3.
```lisp
:pe subsetp ; (+ 1 2)
:pl (+ 1 2)
```

The original regex in the code i.e. `;[^\r\n]*[\r\n]+` does not work with code block 1, possibly because there is no newline at the end of the comment.

The new regex i.e. `;.*$` without `re.MULTILINE` does not work with code block 2, possibly because `$` does not have its intended meaning of end-of-every-line instead of end-of-string.

Code block 3 shows that one does not need to write `;.*?$` i.e. whether the `.*` is greedy or not does not matter.

Why didn't I use `;[^\r\n]*$`? Well, I was in a hurry and used a regex pattern that I knew would work. Sorry about that.

Reply:
Does the above answer your question?

Reply:
Or do you mean something like this?

```python
# Remove all comments from the code block.
cmd = re.sub(r';.*$', '', code, flags=re.MULTILINE)

# Count all commands in the code block. There are 2 types of
# commands recognized:
#
# 1. Keyword commands, such as:
#      :pl subsetp
#    using the ^[ \t]*:.*$ regex. This is a simple recognizer that
#    treats the whole line as the keyword command.
# 2. Top-level LISP forms, such as:
#      (defun app (x y)
#        (cond ((endp x) y)
#              (t (cons (car x)
#                       (app (cdr x) y)))))
#    using the \((?>[^()]|(?R))*\) regex.
num_cmds = len(regex.findall(r'^[ \t]*:.*$|\((?>[^()]|(?R))*\)', cmd, regex.MULTILINE))

# Convert all \r and \n into spaces.
cmd = re.sub(r'[\r\n]', ' ', cmd.strip())
```

Reply:
Yes, that's right. Please do so.

Reply:
Ok no problem 3195bce